src/read/read_options.rs

```diff
@@ -81,6 +84,28 @@ impl ReadOptions {
         self
     }
 
+    /// Guess the file format based on the file name extensions.
+    pub fn guess_format(&mut self, filename: &str) -> &mut Self {
+        let extensions: Vec<&str> = filename.split('.').collect();
+        match extensions[extensions.len() - 1] {
```

Comment:
This will crash on files without any extension. It might be worthwhile to look into using [`Path::extension`](https://doc.rust-lang.org/std/path/struct.Path.html#method.extension) to not have to think about all edge cases that exist.

Reply:
Also I think I saw this same logic match in multiple places (sometimes with and sometimes without the compression) it might be good to refactor this into one function and call that from all places where the logic is needed.